core/src/banking_trace.rs

```diff
@@ -0,0 +1,510 @@
+        if !self.is_checked {
+            self.is_checked = true;
+            self.basic.should_rollover(now, current_filesize)
+        } else {
+            false
+        }
+    }
+}
+
+impl<'a> Write for GroupedWriter<'a> {
+    fn write(&mut self, buf: &[u8]) -> std::result::Result<usize, io::Error> {
+        self.underlying.write_with_datetime(buf, &self.now)
+    }
+    fn flush(&mut self) -> std::result::Result<(), io::Error> {
+        self.underlying.flush()
+    }
+}
+
+pub fn sender_overhead_minimized_receiver_loop<T, E, const SLEEP_MS: u64>(
+    exit: Arc<AtomicBool>,
+    receiver: Receiver<T>,
+    mut on_recv: impl FnMut(T) -> Result<(), E>,
+) -> Result<(), E> {
+    'outer: while !exit.load(Ordering::Relaxed) {
+        'inner: loop {
+            // avoid futex-based blocking here, otherwise a sender would have to
+            // wake me up at a syscall cost...
+            match receiver.try_recv() {
+                Ok(message) => on_recv(message)?,
+                Err(TryRecvError::Empty) => break 'inner,
+                Err(TryRecvError::Disconnected) => {
+                    break 'outer;
+                }
+            };
+            if exit.load(Ordering::Relaxed) {
+                break 'outer;
+            }
+        }
+        sleep(Duration::from_millis(SLEEP_MS));
+    }
+
+    Ok(())
+}
+
+impl BankingTracer {
+    pub fn new(maybe_config: Option<(PathBuf, Arc<AtomicBool>, u64)>) -> Result<Self, TraceError> {
+        let enabled_tracer = maybe_config
+            .map(|(path, exit, total_size)| -> Result<_, TraceError> {
+                let rotate_threshold_size = total_size / TRACE_FILE_ROTATE_COUNT;
+                if rotate_threshold_size == 0 {
+                    return Err(TraceError::TooSmallTraceSize(
+                        total_size,
+                        TRACE_FILE_ROTATE_COUNT,
+                    ));
+                }
+
+                let (trace_sender, trace_receiver) = unbounded();
+
+                Self::ensure_prepare_path(&path)?;
+                let file_appender = Self::create_file_appender(path, rotate_threshold_size)?;
+
+                let tracing_thread =
+                    Self::spawn_background_thread(trace_receiver, file_appender, exit.clone())?;
+
+                Ok((trace_sender, Some(tracing_thread), exit))
+            })
+            .transpose()?;
+
+        Ok(Self { enabled_tracer })
+    }
+
+    pub fn new_disabled() -> Self {
+        Self {
+            enabled_tracer: None,
+        }
+    }
+
+    fn create_channel(&self, label: ChannelLabel) -> (BankingPacketSender, BankingPacketReceiver) {
+        Self::channel(
+            label,
+            self.enabled_tracer
+                .as_ref()
+                .map(|(sender, _, exit)| (sender.clone(), exit.clone())),
+        )
+    }
+
+    pub fn create_channel_non_vote(&self) -> (BankingPacketSender, BankingPacketReceiver) {
```

Comment:
minor design:
It seems odd to me that the tracer is always responsible for creating the channels, even in cases where we don't trace. Seems like banking stage should do that, and the tracer should optionally take/replace the channels.

Reply:
hey yummy :) i like these design talks. let me be a bit chatty here.. ;)

Yeah, I'm aware of the awkwardness around abstraction leakage and misplaced responsibility of channel creations at the first sight as well. However, I intentionally chosen the current design as follows:

as for the awkwardness, `BankingTracer` can conceptually be thought of a middleman between the sigverify stage and the banking stage (maybe, this is tagline comment of `struct BankingTracer`, which i'll add later). i.e. it's not something _purely internal to_ banking stage (Now that it's found that we need `BankingTracer` to sneak into `ReplayStage`  51b553f17bceac10d233d7f9e570a64b29458fa7, I can more say it's not internal to banking stage anymore.).

more specifically, the middleman is a central gate keeper for any incoming inputs into the banking stage (to _possibly_ trace as the name (`BankingTracer`) indicates). So, to observe the subject (banking stage), _it must be outside of the subject in some way for objective observation_. Thus, under that re-interpretation, it's not odd for it to create something through which inputs (= tasks) are flowing from sigverify to banking (i.e. channels). Also, the awkwardness can be further alleviated once  `BankingTracer` is enabled by default in the near future.

Current design can be said to be of [Null object pattern](https://en.wikipedia.org/wiki/Null_object_pattern), and it's not rare to see some uses elsewhere in rust: (1/2) https://doc.rust-lang.org/std/convert/fn.identity.html#examples and (2/2) https://github.com/bincode-org/bincode/blob/c44b5e364e7084cdbabf9f94b63a3c7f32b8fb68/src/config/limit.rs#L39-L49. In our case, likewise, we always create and pass around `TracedSender`s, nevertheless tracing is enabled or not. Thanks to it, we can reduce code-path diversion as much as possible for predictable runtime behavior and minimized risks of bugs. Also, due to rust's ownership, top-down object scattering across subsystems is especially straightforward. And replacing the channels dynamically would difficult to spot in source code wise (vs the current design's ubiquitous `BankingTracer`/`TracedSender` around code base)

Lastly, there's some challenge of the alternative design: BankingStage will take receiver of channels, however we want to fiddle with senders of channels (this is required to record the timings of _task creation_ to benchmark/simulate/investigate, not task reception). So yet again, we just better off setting things up at a one-level higher layer than the banking.

Reply:
>  here's the interface on the tracer I would have imagined:

oh, i didn't considered this runtime interception. i think there's additional channel involved only when traced (batch -> channel created by sigverify -> channel created by enabled BankingTracer -> BankingStage) on top of another channel to feed into tracer's background thread? I'd rather prefer to avoid these additional runtime cost, reducing it to the bare minimum,  if possible. :)

Reply:
(note that such channel daisy chaining requires small yet another thread just to relay as fast as it can, which means syscalls, not just few atomics)

Reply:
Ah, I had a misconception of how this was working; I thought we were already daisy-chaining these. This conversation clears this up:

Sigverify / sender will do a few ops (system_time, arc-clone, send copy to tracer) before sending it to banking stage.
I think this is a much better approach than what I was suggesting!